### ISSUE TITLE
local jqueryui server build

### ISSUE THREAD

**Issue by superyyao** [jquery/download.jqueryui.com#384]
You must use linux  OS to build the service。
jQuery UI DownloadBuilder & ThemeRoller backend and frontend application.

[http://www.xmlsoft.org/downloads.html]()
Requirements
node >= 0.12.x and npm
ImageMagick 6.6.x. (see below for instructions to compile it from source)
grunt-cli: npm install -g grunt-cli
api.jqueryui.com requirements.

**ImageMagick installation to see the official documentation Appendix**

1. 

- install git [https://git-scm.com/download/linux](url)

- apt-get install git
- sudo  ln -s  /home/ubuntu/node-v10.15.3-linux-x64(you node directory)/bin/grunt /usr/local/bin/

- npm install -g bower

- sudo  ln -s  /home/ubuntu/node-v10.15.3-linux-x64(you node directory)/bin/bower /usr/local/bin/

2.  
download.jqueryui.com-master [https://github.com/jquery/download.jqueryui.com](url)
npm install
grunt prepare
(if "grunt prepare" so slow，cd /home/ubuntu/download.jqueryui.com-master/tmp
 git clone https://github.com/jquery/api.jquery.com.git
 git clone https://github.com/jquery/jquery-ui.git
modify Gruntfile.js
```
function( callback ) {
			//grunt.util.spawn({
					//cmd: "git",
					//args: [ "fetch" ],
					//opts: {
						//cwd: "tmp/jquery-ui"
					//}
					//}, callback );
                          callback ();
			},
```

```
function( callback ) {
		//grunt.util.spawn({
		//cmd: "git",
	       //args: [ "fetch", "-t" ],
	      //opts: {
	//cwd: "tmp/jquery-ui"
	//}
	//}, callback );
          callback ();
	}
```
```
	grunt.util.spawn({
		cmd: "bower",
		args: [ "install", "--allow-root" ]
	}, done );
```
)

3. node server.js --console
  if use virtual machine
modify Gruntfile.js
```
server.listen( httpPort, function() {
	console.log()
	console.log( "HTTP Server running at http://%s:%d", httpHost, httpPort );
});
	}, done );
```

good luck!

**Comment by minj** [comment]
instructions unclear lol

e.g. formatting is messed up

> if use virtual machine
> modify Gruntfile.js
> ```
> server.listen( httpPort, function() {
> 	console.log()
> 	console.log( "HTTP Server running at http://%s:%d", httpHost, httpPort );
> });
> 	}, done );
> ```

which, where, what?

@superyyao why not make a PR with a proper diff?

**Comment by superyyao** [comment]
Sorry，I don't quite follow you.
This setup is mainly for me to access the virtual machine's server from the local computer.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;"LA-MJ"<notifications@github.com&gt;;
发送时间:&nbsp;2019年12月18日(星期三) 晚上10:21
收件人:&nbsp;"jquery/download.jqueryui.com"<download.jqueryui.com@noreply.github.com&gt;;
抄送:&nbsp;"yy.yin"<250351662@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [jquery/download.jqueryui.com] local jqueryui server build (#384)




instructions unclear lol
 
e.g. formatting is messed up
  
if use virtual machine
 modify Gruntfile.js
 server.listen( httpPort, function() { 	console.log() 	console.log( "HTTP Server running at http://%s:%d", httpHost, httpPort ); }); 	}, done );   
which, where, what?
 
@superyyao why not make a PR with a proper diff?

**Comment by superyyao** [comment]
Sorry，I don't quite follow you.
This setup is mainly for me to access the virtual machine's server from the local computer

**Comment by superyyao** [comment]
Sorry，I don't quite follow you.
This setup is mainly for me to access the virtual machine's server from the local computer
@minj

**Comment by superyyao** [comment]
Sorry , **modify server.js** not Gruntfile.js
`
`if use virtual machine
**modify server.js**

server.listen( httpPort, function() {
	console.log()
	console.log( "HTTP Server running at http://%s:%d", httpHost, httpPort );
});
	}, done );`
Thanks, @minj

**Comment by minj** [comment]
yeah, my problem is I can't get an older build than 1.12.1 to work. It just times out.

I'll try the VM, thanks.

**Comment by fnagel** [comment]
Related to #359

**Comment by astehlik** [comment]
Hi there,

I created a docker image that allows you to run the jQuery UI theme roller and download builder locally:

https://hub.docker.com/repository/docker/intera/jquery-ui

I hope it helps to run the Theme Roller locally.

It should be as simple as running

```bash
docker run --rm -p 8088:8088 intera/jquery-ui:latest
```

and the opening this URL in your browser:

http://localhost:8088/

**Comment by japicoder** [comment]
@astehlik your solution worked pretty well for me, no overhead and easy to get it running. Thank you!

**Comment by ddison** [comment]
@astehlik Your solution did work great but I am only able to download 1.12 themes it appears and I need 1.9.x themes for the app I have. Any ideas for a workaround?

**Comment by astehlik** [comment]
@ddison Thanks for your feedback.

I tried to download the 1.9.2 version and also got an error. After some digging I was able to solve the problem with a patch. Please pull the newest Docker image and try again.

If you are interested: the patch applied can be found here:
https://github.com/Intera/docker-jquery-ui/blob/master/patches/fix-basename-calls.diff

I hope it works with the new image.

**Comment by rubo77** [comment]
Great! Maybe someone likes to run the docker image on their server so everybody can access it to download the latest builds

**Comment by gbjbaanb** [comment]
> 
> 
> Great! Maybe someone likes to run the docker image on their server so everybody can access it to download the latest builds

maybe, even, I don't know, possibly someone like, perhaps, the OpenJS team who broke the download in the first place :)

> 
> 
> Hi there,
> 
> I created a docker image that allows you to run the jQuery UI theme roller and download builder locally:
> 
> https://hub.docker.com/repository/docker/intera/jquery-ui
> 
> I hope it helps to run the Theme Roller locally.
> 
> It should be as simple as running
> 
> ```shell
> docker run --rm -p 8088:8088 intera/jquery-ui:latest
> ```
> 
> and the opening this URL in your browser:
> 
> http://localhost:8088/

@jorydotcom would it not be possible to run this dockerimage on your infrastructure, maybe with a link from the old download page, so that the builder would be available to everyone unable (or unwilling) to run docker images.

**Comment by astehlik** [comment]
@gbjbaanb I like the idea and I'll see if I can setup a public server.

Unfortunately there is an issue when loading an existing theme that needs to be sorted out first:

https://github.com/Intera/docker-jquery-ui/issues/1

Without fixing that running a public server would not make much sense ;) Any help on this is highly appreciated.

**Comment by astehlik** [comment]
A public demo of the Docker image is now running here:

**Theme Roller**

https://jquery-ui.intera.ws/themeroller/

**Download Builder**

https://jquery-ui.intera.ws/download/

**Comment by mgol** [comment]
Hey, I've just merged my PR #552 & I deployed it; it has a chance to fix those 502 errors. Now, I don't want to give any guarantees since last time I thought it was fixed it broke within an hour but for now it seems to work. Please test & report issues if you find something is still broken.

This is tracked in #390 so we can continue the discussion there.

**Comment by fnagel** [comment]
Here is a working Docker image based upon node-10 including all needed fixes: https://github.com/fnagel/docker-jqueryui

Closing this ticket now, please continue in #390
